swiftsimio/accelerated.py

```diff
@@ -168,3 +168,317 @@ def index_dataset(handle: Dataset, mask_array: np.array) -> np.array:
+
+
+    Parameters
+    ----------
+    ranges : np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+
+    Returns
+    -------
+    np.ndarray
+        two dimensional array of ranges
+
+    Examples
+    --------
+    >>> concatenate_ranges([[1,5],[6,10],[12,15]])
+    np.ndarray([[1,10],[12,15]])
+    """
+    concatenated = [ranges[0]]
+
+    for i in range(1, len(ranges)):
+        lower = ranges[i][0]
+        upper = ranges[i][1]
+        if lower <= concatenated[-1][1] + 1:
+            concatenated[-1][1] = upper
+        else:
+            concatenated.append(ranges[i])
+
+    return np.array(concatenated)
+
+
+@jit(nopython=True, fastmath=True)
+def get_chunk_ranges(
+    ranges: np.ndarray, chunk_size: np.ndarray, array_length: int
+) -> np.ndarray:
+    """
+    Return indices indicating which hdf5 chunk each range from `ranges` belongs to
+
+    Parameters
+    ----------
+    ranges : np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+    chunk_size : int
+        size of the hdf5 dataset chunks
+    array_length: int
+        size of the dataset
+
+    Returns
+    -------
+    np.ndarray
+        two dimensional array of bounds for the chunks that contain each range from
+        `ranges`
+
+    """
+    chunk_ranges = []
+    n_ranges = len(ranges)
+    for bounds in ranges:
+        lower = (bounds[0] // chunk_size) * chunk_size
+        upper = min(-((-bounds[1]) // chunk_size) * chunk_size, array_length)
+
+        # Before appending new chunk range we need to check
+        # that it doesn't already exist or overlap with an
+        # existing one. The only way overlap can happen is
+        # if the current lower index is less than or equal
+        # to the previous upper one. In that case simply
+        # update the previous upper to cover current chunk
+        if len(chunk_ranges) > 0:
+            if lower > chunk_ranges[-1][1]:
+                chunk_ranges.append([lower, upper])
+            elif lower <= chunk_ranges[-1][1]:
+                chunk_ranges[-1][1] = upper
+        # If chunk_ranges is empty, don't do any checks
+        else:
+            chunk_ranges.append([lower, upper])
+
+    return np.array(chunk_ranges)
+
+
+@jit(nopython=True, fastmath=True)
+def expand_ranges(ranges: np.ndarray) -> np.array:
+    """
+    Return an array of indices that are within the specified ranges
+
+    Parameters
+    ----------
+    ranges : np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+
+    Returns
+    -------
+    np.array
+        1D array of indices that fall within each range specified in `ranges`
+        
+    """
+    length = np.asarray([bounds[1] - bounds[0] for bounds in ranges]).sum()
+
+    output = np.zeros(length, dtype=np.int64)
+    i = 0
+    for bounds in ranges:
+        lower = bounds[0]
+        upper = bounds[1]
+        bound_length = upper - lower
+        output[i : i + bound_length] = np.arange(lower, upper, dtype=np.int64)
```

Comment:
Please can you replace this line with the explicit loop that I have requested before merge? It should not be slower if we jit the function and I really do not like this version. Sorry for nit picking.